File: tests/fixtures/open_redirect/go/safe_redirect.go
package handler

import (
	"net/http"
	"strings"
)

// validateRedirectUrl is a project-local allowlist helper: it requires a
// leading `/` to limit redirects to same-origin paths.  Registered as a
// Sanitizer(OPEN_REDIRECT) by `labels/go.rs`.
func validateRedirectUrl(raw string) string {
	if strings.HasPrefix(raw, "/") {
		return raw
	}
	return "/"
}

// Safe: query arg routed through validateRedirectUrl allowlist.
func Safe(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("next")
	safe := validateRedirectUrl(target)
	http.Redirect(w, r, safe, http.StatusFound)
}

File: tests/fixtures/open_redirect/python/safe_redirect.py
# Safe: query arg routed through `validate_redirect_url` allowlist before
# passing to flask.redirect.
from flask import request, redirect


def validate_redirect_url(raw):
    return raw if raw.startswith("/") else "/"


def handler():
    target = request.args.get("next")
    safe = validate_redirect_url(target)
    return redirect(safe)
